File: src/main/java/me/redepicness/bingo/world/WorldManager.java
package me.redepicness.bingo.world;

import me.redepicness.bingo.AnotherBingoPlugin;
import net.kyori.adventure.util.TriState;
import org.bukkit.Bukkit;
import org.bukkit.NamespacedKey;
import org.bukkit.World;
import org.bukkit.World.Environment;
import org.bukkit.WorldCreator;
import org.bukkit.WorldType;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitTask;
import org.intellij.lang.annotations.Subst;
import org.jetbrains.annotations.ApiStatus.Internal;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashSet;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;

public class WorldManager {

    //TODO Make these configurable
    static final int CHUNK_RADIUS_LOADING_STEP = 5;
    static final int CHUNK_LOADS_PER_BATCH = 200;
    static final int MIN_WORLDS_TO_KEEP_READY = 1;
    static final CompletableFuture<?>[] EMPTY_FUTURES_ARRAY = new CompletableFuture[0];

    private final @NotNull Logger logger = LoggerFactory.getLogger("world-manager");
    private final @NotNull AnotherBingoPlugin plugin;
    private final @NotNull Map<NamespacedKey, BingoWorlds> gameWorlds = new ConcurrentHashMap<>();
    private BukkitTask task;

    public WorldManager(@NotNull AnotherBingoPlugin plugin) {
        this.plugin = plugin;
    }

    @Internal
    public void setup(){
        AnotherBingoPlugin plugin = JavaPlugin.getPlugin(AnotherBingoPlugin.class);
        Bukkit.getPluginManager().registerEvents(new WorldListener(), plugin);
        task = Bukkit.getScheduler().runTaskTimerAsynchronously(plugin, this::processWorlds, 0, 10 * 20);
    }

    @Internal
    public void shutdown(){
        if(task != null){
            task.cancel();
        }
        Set<CompletableFuture<Void>> futures = new HashSet<>();
        for (BingoWorlds worlds : new HashSet<>(gameWorlds.values())) {
            futures.add(removeWorlds(worlds));
        }
        CompletableFuture.allOf(futures.toArray(WorldManager.EMPTY_FUTURES_ARRAY)).join();
    }

    public @NotNull Set<BingoWorlds> getAllWorlds(){
        return new HashSet<>(gameWorlds.values());
    }

    public @NotNull Optional<BingoWorlds> getWorlds(@NotNull World world){
        return getWorlds(world.getKey());
    }

    public @NotNull Optional<BingoWorlds> getWorlds(@NotNull NamespacedKey key){
        return Optional.ofNullable(gameWorlds.get(key));
    }

    public @NotNull CompletableFuture<Void> removeWorlds(@NotNull BingoWorlds worlds){
        gameWorlds.remove(worlds.overworld.getKey());
        gameWorlds.remove(worlds.nether.getKey());
        gameWorlds.remove(worlds.end.getKey());
        return worlds.destroyWorlds();
    }

    public @NotNull CompletableFuture<BingoWorlds> createNewWorld(@NotNull BingoWorldsConfig config){
        NamespacedKey overworldKey = generateWorldKey();
        NamespacedKey netherKey = new NamespacedKey(overworldKey.namespace(), overworldKey.getKey()+"-nether");
        NamespacedKey endKey = new NamespacedKey(overworldKey.namespace(), overworldKey.getKey()+"-end");

        CompletableFuture<BingoWorlds> future = new CompletableFuture<>();

        logger.info("Creating new worlds for key {}...", overworldKey);

        WorldCreator overworldCreator = new WorldCreator(overworldKey)
                .environment(Environment.NORMAL)
                .keepSpawnLoaded(TriState.FALSE)
                .type(WorldType.NORMAL);

        WorldCreator netherCreator = new WorldCreator(netherKey)
                .environment(Environment.NETHER)
                .keepSpawnLoaded(TriState.FALSE)
                .type(WorldType.NORMAL);

        WorldCreator endCreator = new WorldCreator(endKey)
                .environment(Environment.THE_END)
                .keepSpawnLoaded(TriState.FALSE)
                .type(WorldType.NORMAL);

        Bukkit.getScheduler().runTaskLater(plugin, () -> {
            logger.info("Creating overworld with key {}...", overworldKey);
            World overworld = overworldCreator.createWorld();
            Bukkit.getScheduler().runTaskLater(plugin, () -> {
                logger.info("Creating nether with key {}...", netherKey);
                World nether = netherCreator.createWorld();
                Bukkit.getScheduler().runTaskLater(plugin, () -> {
                    logger.info("Creating end with key {}...", endKey);
                    World end = endCreator.createWorld();

                    if(overworld == null){
                        logger.warn("Failed to create overworld {}!", overworldKey);
                        future.completeExceptionally(new IllegalStateException());
                        return;
                    }
                    if(nether == null){
                        logger.warn("Failed to create nether {}!", netherKey);
                        future.completeExceptionally(new IllegalStateException());
                        return;
                    }
                    if(end == null){
                        logger.warn("Failed to create end {}!", endKey);
                        future.completeExceptionally(new IllegalStateException());
                        return;
                    }

                    BingoWorlds bingoWorlds = new BingoWorlds(config, overworld, nether, end);
                    gameWorlds.put(overworldKey, bingoWorlds);
                    gameWorlds.put(netherKey, bingoWorlds);
                    gameWorlds.put(endKey, bingoWorlds);

                    logger.info("Created worlds for key {}!", overworldKey);

                    if(config.preloading().preload()){
                        bingoWorlds.startLoadingWorlds();
                    }

                    future.complete(bingoWorlds);

                }, 20);
            }, 20);
        }, 20);

        return future;
    }

    private void processWorlds(){
        int readyWorlds = 0;

        for (BingoWorlds worlds : new HashSet<>(gameWorlds.values())) {
            if(!worlds.used) {
                readyWorlds++;
            }
        }

        for (int i = readyWorlds; i < MIN_WORLDS_TO_KEEP_READY; i++) {
            logger.info("Creating new world, as current number of ready worlds is too low! {} < {}", readyWorlds, MIN_WORLDS_TO_KEEP_READY);
            Bukkit.getScheduler().runTask(plugin, () -> createNewWorld(new BingoWorldsConfig())); //TODO Use a default config
        }
    }

    private @NotNull NamespacedKey generateWorldKey(){
        NamespacedKey key;
        do {
            @Subst("1a2b3c4d") String uuid = UUID.randomUUID().toString().substring(0, 8);
            key = new NamespacedKey("abp", "bingo-"+ uuid);
        }
        while (gameWorlds.containsKey(key));
        return key;
    }

}

File: src/main/kotlin/me/redepicness/bingo/world/WorldManager.kt
package me.redepicness.bingo.world

import me.redepicness.bingo.util.registerEvents
import me.redepicness.bingo.util.runTask
import me.redepicness.bingo.util.runTaskLater
import me.redepicness.bingo.util.runTaskTimerAsynchronously
import net.kyori.adventure.util.TriState
import org.bukkit.NamespacedKey
import org.bukkit.World
import org.bukkit.WorldCreator
import org.bukkit.WorldType
import org.bukkit.scheduler.BukkitTask
import org.intellij.lang.annotations.Subst
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap

object WorldManager {

    //TODO Make these configurable
    const val CHUNK_RADIUS_LOADING_STEP = 5
    const val CHUNK_LOADS_PER_BATCH = 50
    const val MIN_WORLDS_TO_KEEP_READY = 1

    private val logger = LoggerFactory.getLogger("world-manager")
    private val gameWorlds: MutableMap<NamespacedKey, BingoWorlds> = ConcurrentHashMap<NamespacedKey, BingoWorlds>()

    private lateinit var task: BukkitTask
    val allWorlds: Set<BingoWorlds>
        get() = gameWorlds.values.toSet()
    fun setup() {
        registerEvents(WorldListener())
        task = runTaskTimerAsynchronously(0, 10 * 20) { processWorlds() }
    }

    fun shutdown() {
        task.cancel()
        val futures: MutableSet<CompletableFuture<Void>> = mutableSetOf()
        for (worlds in gameWorlds.values.toSet()) {
            futures.add(removeWorlds(worlds))
        }
        CompletableFuture.allOf(*futures.toTypedArray()).join()
    }

    fun getWorldsForGame(): CompletableFuture<BingoWorlds> {
        return allWorlds.find { !it.isUsed }.let { if(it != null) CompletableFuture.completedFuture(it) else createNewWorld(BingoWorldsConfig()) }
    }

    fun getWorlds(world: World) = getWorlds(world.key)

    fun getWorlds(key: NamespacedKey) = gameWorlds[key]

    fun removeWorlds(worlds: BingoWorlds): CompletableFuture<Void> {
        gameWorlds.remove(worlds.overworld.key)
        gameWorlds.remove(worlds.nether.key)
        gameWorlds.remove(worlds.end.key)
        return worlds.destroyWorlds()
    }

    fun createNewWorld(config: BingoWorldsConfig): CompletableFuture<BingoWorlds> {
        val overworldKey: NamespacedKey = generateWorldKey()
        val netherKey = NamespacedKey(overworldKey.namespace, overworldKey.key + "-nether")
        val endKey = NamespacedKey(overworldKey.namespace, overworldKey.key + "-end")
        val future: CompletableFuture<BingoWorlds> = CompletableFuture<BingoWorlds>()
        logger.info("Creating new worlds for key {}...", overworldKey)

        val overworldCreator: WorldCreator = WorldCreator(overworldKey)
                .environment(World.Environment.NORMAL)
                .keepSpawnLoaded(TriState.FALSE)
                .type(WorldType.NORMAL)

        val netherCreator: WorldCreator = WorldCreator(netherKey)
                .environment(World.Environment.NETHER)
                .keepSpawnLoaded(TriState.FALSE)
                .type(WorldType.NORMAL)

        val endCreator: WorldCreator = WorldCreator(endKey)
                .environment(World.Environment.THE_END)
                .keepSpawnLoaded(TriState.FALSE)
                .type(WorldType.NORMAL)

        runTaskLater(20) {
            logger.info("Creating overworld with key {}...", overworldKey)
            val overworld = overworldCreator.createWorld()
            runTaskLater(20) {
                logger.info("Creating nether with key {}...", netherKey)
                val nether = netherCreator.createWorld()
                runTaskLater(20) {
                    logger.info("Creating end with key {}...", endKey)
                    val end = endCreator.createWorld()
                    if (overworld == null) {
                        logger.warn("Failed to create overworld {}!", overworldKey)
                        future.completeExceptionally(IllegalStateException())
                    }
                    else if (nether == null) {
                        logger.warn("Failed to create nether {}!", netherKey)
                        future.completeExceptionally(IllegalStateException())
                    }
                    else if (end == null) {
                        logger.warn("Failed to create end {}!", endKey)
                        future.completeExceptionally(IllegalStateException())
                    }
                    else {
                        val bingoWorlds = BingoWorlds(config, overworld, nether, end)
                        gameWorlds[overworldKey] = bingoWorlds
                        gameWorlds[netherKey] = bingoWorlds
                        gameWorlds[endKey] = bingoWorlds
                        logger.info("Created worlds for key {}!", overworldKey)
                        if (config.preloading.preload) {
                            bingoWorlds.startLoadingWorlds()
                        }
                        future.complete(bingoWorlds)
                    }
                }
            }
        }
        return future
    }

    private fun processWorlds() {
        var readyWorlds = 0
        for (worlds in gameWorlds.values.toMutableSet()) {
            if (!worlds.isUsed) {
                readyWorlds++
            }
        }
        for (i in readyWorlds until MIN_WORLDS_TO_KEEP_READY) {
            logger.info("Creating new world, as current number of ready worlds is too low! {} < {}", readyWorlds, MIN_WORLDS_TO_KEEP_READY)
            runTask { createNewWorld(BingoWorldsConfig()) } //TODO Use a default config
        }
    }

    private fun generateWorldKey(): NamespacedKey {
        var key: NamespacedKey
        do {
            @Subst("1a2b3c4d") val uuid: String = UUID.randomUUID().toString().substring(0, 8)
            key = NamespacedKey("abp", "bingo-$uuid")
        } while (gameWorlds.containsKey(key))
        return key
    }
}
